src/aktualizr_secondary/uptane_verification_test.cc

```diff
@@ -95,104 +195,209 @@ class SecondaryUptaneVerificationTest : public ::testing::Test {
+    auto rev_to_apply = rev.empty() ? _treehub->curRev() : rev;
+    auto hw_id = hardware_id.empty() ? _secondary.hardwareID() : hardware_id;
+    auto serial_id = serial.empty() ? _secondary.serial() : serial;
+
+    _uptane_repo.addOstreeRev(rev, hw_id, serial_id);
 
-    EXPECT_FALSE(_secondary->putMetadataResp(Metadata(malformed_metadata)));
+    return currentMetadata();
   }
 
-  {
-    auto malformed_metadata = _uptane_repo.getCurrentMetadata();
-    malformed_metadata.image_targets[4] = 'f';
+  Metadata currentMetadata() const { return _uptane_repo.getCurrentMetadata(); }
 
-    EXPECT_FALSE(_secondary->putMetadataResp(Metadata(malformed_metadata)));
+  std::shared_ptr<std::string> getCredsToSend() const {
+    std::map<std::string, std::string> creds_map = {
+        {"ca.pem", ""}, {"client.pem", ""}, {"pkey.pem", ""}, {"server.url", _treehub->url()}};
+
+    std::stringstream creads_strstream;
+    Utils::writeArchive(creds_map, creads_strstream);
+
+    return std::make_shared<std::string>(creads_strstream.str());
   }
+
+  Uptane::Hash treehubCurRev() const { return Uptane::Hash(Uptane::Hash::Type::kSha256, _treehub->curRev()); }
+
+ protected:
+  static std::shared_ptr<Treehub> _treehub;
+  static std::string _ostree_rootfs_template;
+  static std::shared_ptr<OstreeRootfs> _sysroot;
+
+  AktualizrSecondaryWrapper _secondary{*_sysroot, *_treehub};
+  UptaneRepoWrapper _uptane_repo;
+};
+
+std::shared_ptr<Treehub> OstreeSecondaryUptaneVerificationTest::_treehub{nullptr};
+std::string OstreeSecondaryUptaneVerificationTest::_ostree_rootfs_template{"./build/ostree_repo"};
+std::shared_ptr<OstreeRootfs> OstreeSecondaryUptaneVerificationTest::_sysroot{nullptr};
+
+TEST_F(OstreeSecondaryUptaneVerificationTest, fullUptaneVerificationPositive) {
+  EXPECT_TRUE(_secondary->putMetadataResp(addDefaultTarget()));
+  EXPECT_TRUE(_secondary->sendFirmwareResp(getCredsToSend()));
+  EXPECT_TRUE(_secondary.getPendingVersion().MatchHash(treehubCurRev()));
+  // TODO: emulate reboot and check installed version once ostree update finalization is supported by secondary
+}
+
+TEST_F(OstreeSecondaryUptaneVerificationTest, fullUptaneVerificationInvalidRevision) {
+  EXPECT_TRUE(_secondary->putMetadataResp(addTarget("invalid-revision")));
+  EXPECT_FALSE(_secondary->sendFirmwareResp(getCredsToSend()));
+}
+
+TEST_F(OstreeSecondaryUptaneVerificationTest, fullUptaneVerificationInvalidHwID) {
+  EXPECT_FALSE(_secondary->putMetadataResp(addTarget("", "invalid-hardware-id", "")));
+}
+
+TEST_F(OstreeSecondaryUptaneVerificationTest, fullUptaneVerificationInvalidSerial) {
+  EXPECT_FALSE(_secondary->putMetadataResp(addTarget("", "", "invalid-serial-id")));
+}
+
+TEST_F(SecondaryUptaneVerificationTest, fullUptaneVerificationPositive) {
+  EXPECT_TRUE(_secondary->putMetadataResp(_uptane_repo.getCurrentMetadata()));
+  EXPECT_TRUE(_secondary->sendFirmwareResp(getImageData()));
 }
 
+TEST_P(SecondaryUptaneVerificationTestNegative, MalformedMetadaJson) {
```

Comment:
Neat! Could you group the TEST_P and INSTANTIATE_TEST_SUITE_P together with some explanatory comments, to make it clearer in further reviews?

Here, the lone `TEST_P` can be easily confused with the series of TEST_F before.

Reply:
ok.